New terrain system

AnsisMalins opened the pull request.

AnsisMalins committed:
Apache license, same as unity-explorer

AnsisMalins committed:
Draw instanced terrain parcels

AnsisMalins committed:
Shader that replicates a subset of what the Decentraland terrain generator does

AnsisMalins committed:
Update gitignore

AnsisMalins committed:
Get the shader to work

AnsisMalins committed:
Played around with noise

AnsisMalins committed:
Generate terrain collision

AnsisMalins committed:
More correct collision creation radius

AnsisMalins committed:
Simplify parcel collider reuse

AnsisMalins committed:
Add basic frustum culling

AnsisMalins committed:
Move common data to a scriptable object

GBirch33 committed:
Start of replacement of Mountain Shader

AnsisMalins committed:
Added trees

AnsisMalins committed:
Extract tree colliders

AnsisMalins committed:
Tree colliders match their renderers

AnsisMalins committed:
Fix frustum culling

AnsisMalins committed:
Scatter grass, also crlf

AnsisMalins committed:
Add Stylized Grass Shader

GBirch33 committed:
rolling hills

GBirch33 committed:
value name change for the material

AnsisMalins committed:
Refactored noise code to be C#/HLSL polyglot

AnsisMalins committed:
Switch to Geoff noise

GBirch33 committed:
derivatives working and new height generation algorithm which is closer to Quilez's

GBirch33 committed:
Shadow fix

AnsisMalins committed:
Jobify the terrain renderer

AnsisMalins committed:
Unity keeps modifying these files

AnsisMalins committed:
Parallelize parcel generation better

AnsisMalins committed:
Terrain LOD WIP

GBirch33 committed:
Terrain optimisation

AnsisMalins committed:
Better frustum culling

AnsisMalins committed:
Clip drawing region to the terrain bounds

GBirch33 committed:
Optimised terrain algorithm with integer based hashing, and forced full floating point over half floating point

AnsisMalins committed:
Apply ground mesh LOD and culling to terrain renderer

AnsisMalins committed:
Move noise code around

AnsisMalins committed:
Fix crashes due to running out of list space

AnsisMalins committed:
Create occupancy texture from world manifest

AnsisMalins committed:
Allow turning off rendering of trees and detail

AnsisMalins committed:
Amend the magic pattern to deal with ground mesh seams

AnsisMalins committed:
Get rid of seams between ground meshes

AnsisMalins committed:
Turn off compression for the ground meshes

AnsisMalins committed:
Update to Unity 6.0.49f1 to match Explorer

AnsisMalins committed:
Clip ground meshes to terrain bounds

AnsisMalins committed:
Implement blending to occupied parcels

AnsisMalins committed:
Amend

AnsisMalins committed:
Have the terrain correctly blend to zero at edges

AnsisMalins committed:
Fix tree placement

AnsisMalins committed:
Turn trees and detail back on

AnsisMalins committed:
Turn on Burst

AnsisMalins committed:
Delete unneeded things

AnsisMalins committed:
Don't place trees and grass on occupied parcels

AnsisMalins committed:
Made a build

AnsisMalins committed:
Made a package

GBirch33 committed:
Add in the Height Texture generating code and generated resources

GBirch33 committed:
Shader changes

AnsisMalins committed:
Adjust the code so it can be integrated into the Decentraland client

AnsisMalins committed:
crlf

mikhail-dcl reviewed (changes requested):
The code looks very carefully structured, good job 👍 
I simply have questions and suggestions directed to possible further improvements

AnsisMalins committed:
Use function pointers for noise

AnsisMalins commented on Packages/org.decentraland.terrain/Runtime/TerrainCollider.cs:
```diff
@@ -0,0 +1,256 @@
+        private void GenerateTree(Vector2Int parcel, in TerrainDataData terrainData, ref Random random,
+            ParcelData parcelData)
+        {
+            if (!terrainData.NextTree(int2(parcel.x, parcel.y), ref random,
```
There is not much to reuse. Collision needs only up to 4 parcels, and each parcel has like 15% chance to spawn a tree.

AnsisMalins commented on Packages/org.decentraland.terrain/Runtime/ParallelWriterExtensions.cs:
```diff
@@ -0,0 +1,25 @@
+{
+    public static class ParallelWriterExtensions
+    {
+        public static unsafe bool TryAddNoResize<T>(this NativeList<T>.ParallelWriter writer, T value)
```
How do you imagine deciding to which array index to write to?

mikhail-dcl replied:
Agree, only that is a limitation

AnsisMalins commented on Packages/org.decentraland.terrain/Runtime/TerrainRenderer.cs:
```diff
@@ -0,0 +1,933 @@
+        internal int TreeInstanceCount { get; private set; }
+#endif
+
+        private void Update()
```
Yes. Coming soon.

AnsisMalins commented on Packages/org.decentraland.terrain/Runtime/TerrainRenderer.cs:
```diff
@@ -0,0 +1,933 @@
+namespace Decentraland.Terrain
+{
+    [ExecuteAlways]
+    public sealed class TerrainRenderer : MonoBehaviour
```
They are visible in the profiler already?

mikhail-dcl replied:
If so ignore this comment

AnsisMalins commented on Packages/org.decentraland.terrain/Runtime/TerrainRenderer.cs:
```diff
@@ -0,0 +1,933 @@
+                treeInstanceCounts.Dispose();
+                treeTransforms.Dispose();
+
+                prepareDetailRenderList.Complete();
```
Yes, it would be nice to split the update function into two: StartJobs and FinishJobs, and put them as far apart in the frame as possible. Coming eventually.

AnsisMalins commented on Packages/org.decentraland.terrain/Runtime/TerrainRenderer.cs:
```diff
@@ -0,0 +1,933 @@
+                {
+                    terrainData.detailInstanceCapacity = (int)ceil(terrainData.detailInstanceCapacity * 1.1f);
+
+                    Debug.LogWarning(
```
How do I access ReportHub in a package?

AnsisMalins commented on Packages/org.decentraland.terrain/Runtime/TerrainRenderer.cs:
```diff
@@ -0,0 +1,933 @@
+                    {
+                        renderParams.material = lod.materials[subMeshIndex];
+
+                        Graphics.RenderMeshInstanced(renderParams, lod.mesh, subMeshIndex, instanceData,
```
We're working on it. Regular instanced was easier to do, so that's why I did that first.

AnsisMalins committed:
Use properties with backing fields

AnsisMalins committed:
Button to dump the occupancy map

AnsisMalins committed:
Jobified the terrain collider

AnsisMalins committed:
amend

AnsisMalins committed:
Move terrain collider state out of the component

AnsisMalins committed:
Add user component to the camera in the sample scene

GBirch33 committed:
MountainLit shader updates

GBirch33 committed:
Addition Noise functions

AnsisMalins committed:
Add Geoff terrain data and adjust depth bias

AnsisMalins committed:
Use Geoff terrain data in the sample scene

AnsisMalins committed:
amend

AnsisMalins committed:
Fix Geoff's terrain data

AnsisMalins committed:
Fixed rocks

AnsisMalins committed:
Prevent grass from overlapping with scenes

AnsisMalins committed:
Make trees not overlap with each other and scenes WIP

GBirch33 committed:
Move Terrain to fully utilise texture for height in GPU rendering

GBirch33 committed:
TerrainHeightMap metadata

GBirch33 committed:
Reassigned Terrain Heightmap Texture to material

GBirch33 committed:
GrassIndirectRendering initial setup

GBirch33 committed:
Grass scale, Y rotation and terrain normal alignment

AnsisMalins committed:
Change occupancy map to be power of two WIP

AnsisMalins committed:
Fix collision

GBirch33 committed:
Grass using new occupancy

AnsisMalins committed:
Make indirect grass renderer work in edit mode

AnsisMalins committed:
Make indirect renderer work with no occupancy map

GBirch33 committed:
Safety checks

GBirch33 committed:
Add missing file

GBirch33 committed:
Grass blending from terrain texture

GBirch33 committed:
SatelliteFloor shader basic implementation

GBirch33 committed:
grass blending

GBirch33 committed:
TerrainBounds honoured in Compute Shader

AnsisMalins committed:
Blue noise trees done

AnsisMalins committed:
Fixed an off by one in the grass scatter shader

AnsisMalins committed:
Move compute shaders to the package

AnsisMalins committed:
Fix a crash when terrain area is zero

AnsisMalins committed:
Rocks and bushes rendering

GBirch33 committed:
Addition of _ColorMapsParams value to avoid extra variant

AnsisMalins committed:
Unbreak the build

AnsisMalins committed:
Switched to trees imported from the current implementation

AnsisMalins committed:
Fix an ObjectDisposedException

AnsisMalins committed:
Snap trees to terrain

AnsisMalins committed:
Don't throw when a tree has no collider

GBirch33 committed:
Flowers and Cat tails compute shaders

# Conflicts:
#	Packages/org.decentraland.terrain/Runtime/ComputeCulling/GrassIndirectRenderer.cs
#	Packages/org.decentraland.terrain/Runtime/TerrainData.cs

GBirch33 committed:
add terrain textures

GBirch33 committed:
Terrain error fix

GBirch33 committed:
bringing back terrain to testable settings

GBirch33 committed:
compute adjustments

GBirch33 committed:
Grass and Flower bounds fix

GBirch33 committed:
Terrain Bounds 100 percent fix

AnsisMalins committed:
Split code up to make it more manageable

GBirch33 committed:
Cat tails jittering fix

AnsisMalins committed:
Implement SpeedTree wind

AnsisMalins committed:
Expose some renderer state

GBirch33 committed:
Removed the GPU readbacks

popuz committed:
add DetailDistance to the frustum similar to GrassRendering

popuz committed:
add DetailDistance to the frustum similar to GrassRendering

popuz committed:
Merge remote-tracking branch 'origin/main'

popuz committed:
clamp frustrum far place by min between camera and detail distance

popuz committed:
add debug occupancy

popuz committed:
adjusted Height

popuz committed:
Revert "adjusted Height"

This reverts commit 7ae78220c911efd67a2e54e4057aa2d81af0eacf.

popuz committed:
distance field embedded in occupancy map

popuz committed:
correct saving to 512x512

popuz committed:
adjusted proper calculation

popuz committed:
assigned new texture

popuz committed:
working height

popuz committed:
with step approach

popuz committed:
add smoothness and cleanup Goeff mat

popuz committed:
changes on MountainLit shader for heights

popuz committed:
add debug for occupancy

popuz committed:
adjusted MountainNoise

popuz committed:
- adjusted to inverted map; - simplified without the noise

popuz committed:
adjusted GeoffNoise

GBirch33 committed:
clamp the farclip plane for the grass

popuz committed:
adjusted terrain collider

popuz committed:
set equal height everywhere

popuz committed:
adapted grass to new dist occupancy map texture

popuz committed:
mountain noise working good

popuz committed:
adjusted Geoff noise

popuz committed:
Merge branch 'main' into fix-heightmap

popuz committed:
clean up of assets

popuz committed:
more clean up

popuz committed:
more clean up

popuz committed:
last small clean up

popuz committed:
changed filterMode for distOccup

popuz committed:
Merge pull request #2 from decentraland/fix-heightmap

NOTE: normals calculation are not adapted, so changes in `MountainLitForwardPass.hlsl` are just a placeholder
